amazon/ionhash/hasher.py

```diff
@@ -0,0 +1,462 @@
+    couroutine's behavior matches that of the wrapped writer.
+
+    Args:
+        writer(coroutine):  An ion-python writer coroutine.
+        hash_function_provider(function):  A function that returns a function that produces
+            ``IonHasher`` instances when called.  Note that multiple ``IonHasher`` instances
+            may be required to hash a single value (depending on the type of the Ion value).
+
+    Yields:
+        bytes:  The result of hashing.
+        other values:  As defined by the provided writer coroutine.
+    """
+    hw = _hasher(_hash_writer_handler, writer, hash_function_provider)
+    next(hw)    # prime the coroutine
+    return hw
+
+
+def _hasher(handler, delegate, hash_function_provider):
+    """Provides a coroutine that wraps an ion-python reader or writer and adds Ion Hash functionality."""
+    hasher = _Hasher(hash_function_provider)
+    output = None
+    input = yield output
+    while True:
+        if isinstance(input, HashEvent):
+            if input == HashEvent.DIGEST:
+                input = yield hasher.digest()
+        else:
+            output = handler(input, output, hasher, delegate)
+            if output is None:
+                break
+            input = yield output
+
+
+def _hash_reader_handler(input, output, hasher, reader):
+    """Handles input to a reader-based coroutine."""
+    if isinstance(input, DataEvent):
+        if input == SKIP_EVENT:
+            # translate SKIP_EVENTs into the appropriate number
+            # of NEXT_EVENTs to ensure hash correctness:
+            target_depth = output.depth
+            if output.event_type != IonEventType.CONTAINER_START:
+                target_depth = output.depth - 1
+
+            output = reader.send(NEXT_EVENT)
+            while output.event_type != IonEventType.STREAM_END and output.depth > target_depth:
+                _hash_event(hasher, output)
+                output = reader.send(NEXT_EVENT)
+
+        else:
+            output = reader.send(input)
+
+    if isinstance(output, IonEvent):
+        _hash_event(hasher, output)
+
+    return output
+
+
+def _hash_writer_handler(input, output, hasher, writer):
+    """Handles input to a writer-based coroutine."""
+    if isinstance(input, IonEvent):
+        output = writer.send(input)
+        _hash_event(hasher, input)
+
+    return output
+
+
+def _hash_event(hasher, event):
+    """Maps an IonEvent to the appropriate hasher method."""
+    if event.event_type is IonEventType.CONTAINER_START:
+        hasher.step_in(event)
+    elif event.event_type is IonEventType.CONTAINER_END:
+        hasher.step_out()
+    elif event.event_type is not IonEventType.STREAM_END:
+        hasher.scalar(event)
+
+
+# Ion Hash special bytes:
+_BEGIN_MARKER_BYTE = 0x0B
+_END_MARKER_BYTE = 0x0E
+_ESCAPE_BYTE = 0x0C
+_BEGIN_MARKER = bytes([_BEGIN_MARKER_BYTE])
+_END_MARKER = bytes([_END_MARKER_BYTE])
+
+# Type/Qualifier byte for each Ion type:
+_TQ = {
```

Comment:
For certain types (e.g. INT), this _TQ lookup table is only used for null values, correct? Might be worth a comment. At first I was afraid something was missing as INT has both `0x20` and `0x30`. Then I realized it defers to `_serialize_int` for non-null values.